pkg/handlers/jsonl.go

```diff
@@ -147,112 +148,139 @@ func (j JSONLHandler) handleJSONLObject(f *appendable.IndexFile, r []byte, dec *
+					// nil values are a bit of a degenerate case, we are essentially using the btree
+					// as a set. we store the value as an empty byte slice.
 					if err := page.BPTree(&btree.BPTree{Data: r, DataParser: j, Width: width}).Insert(btree.ReferencedValue{
+						Value:       []byte{},
 						DataPointer: mp,
-						Value:       []byte{0},
 					}, data); err != nil {
-						return fmt.Errorf("failed to insert into b+tree: %w", err)
+						return fmt.Errorf("failed to insert into b+tree: %w\nmp: %v", err, data.Offset)
 					}
-				}
-			case json.Token:
-				switch value {
-				case json.Delim('['):
-					// arrays are not indexed yet because we need to incorporate
-					// subindexing into the specification. however, we have to
-					// skip tokens until we reach the end of the array.
-					depth := 1
-					for {
-						t, err := dec.Token()
+				case appendable.FieldTypeFloat64, appendable.FieldTypeUint64, appendable.FieldTypeInt64:
+					buf := make([]byte, 8)
+					switch value := value.(type) {
+					case json.Number:
+						f, err := value.Float64()
 						if err != nil {
-							return fmt.Errorf("failed to read token: %w", err)
+							return fmt.Errorf("failed to parse float: %w", err)
 						}
+						binary.BigEndian.PutUint64(buf, math.Float64bits(f))
+					case float64:
+						binary.BigEndian.PutUint64(buf, math.Float64bits(value))
+					}
 
-						switch t {
-						case json.Delim('['):
-							depth++
-						case json.Delim(']'):
-							depth--
+					if err := page.BPTree(&btree.BPTree{Data: r, DataParser: j, Width: width}).Insert(btree.ReferencedValue{
+						DataPointer: mp,
+						Value:       buf,
+					},
+						data); err != nil {
+						return fmt.Errorf("failed to insert into b+tree: %w", err)
+					}
+				case appendable.FieldTypeBoolean:
+					valueBool, ok := value.(bool)
+					if !ok {
+						return fmt.Errorf("expected bool type")
+					}
+					if valueBool {
+						if err := page.BPTree(&btree.BPTree{Data: r, DataParser: j, Width: width}).Insert(btree.ReferencedValue{
+							DataPointer: mp,
+							Value:       []byte{1},
+						}, data); err != nil {
+							return fmt.Errorf("failed to insert into b+tree: %w", err)
 						}
-
-						if depth == 0 {
-							break
+					} else {
+						if err := page.BPTree(&btree.BPTree{Data: r, DataParser: j, Width: width}).Insert(btree.ReferencedValue{
+							DataPointer: mp,
+							Value:       []byte{0},
+						}, data); err != nil {
+							return fmt.Errorf("failed to insert into b+tree: %w", err)
 						}
 					}
-				case json.Delim('{'):
-					// find the index to set the field type to unknown.
-					if err := j.handleJSONLObject(f, r, dec, append(path, key), data); err != nil {
-						return fmt.Errorf("failed to handle object: %w", err)
-					}
-					// read the }
-					if t, err := dec.Token(); err != nil || t != json.Delim('}') {
-						return fmt.Errorf("expected '}', got '%v'", t)
+				case appendable.FieldTypeArray, appendable.FieldTypeObject:
```

Comment:
since these are degenerate cases (*as of now*), I hope it's alright to case them together and combine the logic

Reply:
Yep that's fine.